### ISSUE TITLE
yum state=latest doesn't install if not already installed

### ISSUE THREAD

**Issue by jkleint** [ansible/ansible#269]
If I try to install a package which is not already installed using yum with state=latest:

```
ansible -m yum -a 'pkg=hadoop-hdfs-namenode state=latest' box -Ksu$USER
```

It does a `yum update`, which (stupidly, if you ask me) doesn't install the package.  However, yum returns 0, so ansible thinks it was installed.

There is some logic in the yum module that tries to do a `yum install` if the package isn't installed.  This line:

```
    updates = my.doPackageLists(pkgnarrow='updates', patterns=[pkgspec]).updates
```

checks first to see if an updated version of the package is available.  However, on my systems (rhel5, yum 3.2.22), narrowing down the `patterns` doesn't seem to be working, and it just returns the list of whatever security updates are available from RedHat, which does not include the package I care about.

This line:

```
    avail =  my.doPackageLists(pkgnarrow='available', patterns=[pkgspec]).available
```

Does contain only the package I'm trying to install.  But then the logic goes

```
    if not updates:
        cmd = "yum -c %s -d1 -y install '%s'" % (yumconf, pkgspec)
    else:
        cmd = "yum -c %s -d1 -y update '%s'" % (yumconf, pkgspec)
```

So it takes the `update` path, which doesn't do anything.

I don't know if that's a bug or a feature of yum, and whether it's fixed in later versions, so I'm not really sure what the fix is.

**Comment by mpdehaan** [comment]
assigning to Seth, but this should work regardless of how yum operates, because that is the intent.

I don't recall seeing this happen to me though, which is strange.

**Comment by skvidal** [comment]
not sure how many options we're going to have on rhel5 with that version of yum.

I'll test on my rhel5 box here and let you know.

**Comment by mpdehaan** [comment]
seems like in the worst case we figure out that it failed to install from the output and then do the right thing.

It just needs to switch to "yum install" if that fails in that way, right?

**Comment by skvidal** [comment]
on my centos 5 boxes with yum 3.2.22-39 I cannot replicate this:

mycentos5$ rpm -q zsh
package zsh is not installed

ansible --timeout=30 -i ~/alist mycentos5 -m yum -a "pkg=zsh state=latest" 

mycentos5 | success >> {
    "changed": true
}

mycentos5$ rpm -q zsh
zsh-4.2.6-6.el5

Can you provide more debugging information?

**Comment by jkleint** [comment]
Yes.  I'll bet you have all your security updates installed, don't you?  That is, `sudo yum check-update` shows no packages in need of updating?  In this case, the Python `updates` list is empty, and you do an install.  However, I have not done my security updates (shame on me!), and so that list comes back with all the packages I need to update (and not the one I'm trying to install).  In other words, the `patterns=[pkgspec]` isn't working (or RedHat doesn't want it to).  So I take the update path, which doesn't do anything.

I verified this by installing all updates on one machine; `state=latest` then works properly.  So, taking the charitable view, I suppose ansible is really encouraging me to update?  ;)

**Comment by mpdehaan** [comment]
jkleint -- please refrain from "Red Hat doesn't want it to" if you want any support around here.   Think about who wrote some of your code.

**Comment by skvidal** [comment]
okay - I figured out how to replicate this case.

doPackageLists in rhel5's yum doesn't have the patterns= limiting support that following versions did have.

It's been a long time since it mattered to me so I didn't notice it.

the yum module here:
https://github.com/skvidal/ansible/blob/devel/library/yum

I've put a pull request in on should solve it. It does in my tests.

I have an alternative fix that involves checking if there are available pkgs matching patterns to pass yum install as the args. Either will work
I chose this path b/c it makes things be more or less the same on rhel5 as rhel6

jkleint: 
 the name of the company is Red Hat, not RedHat
